mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/scheduler/ResourceClusterAwareSchedulerActor.java

```diff
@@ -0,0 +1,318 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.mantisrx.server.master.scheduler;
+
+import static akka.pattern.Patterns.pipe;
+
+import akka.actor.AbstractActorWithTimers;
+import akka.actor.Props;
+import akka.japi.pf.ReceiveBuilder;
+import io.mantisrx.server.core.domain.WorkerId;
+import io.mantisrx.server.master.ExecuteStageRequestFactory;
+import io.mantisrx.server.master.resourcecluster.ResourceCluster;
+import io.mantisrx.server.master.resourcecluster.TaskExecutorID;
+import io.mantisrx.server.master.resourcecluster.TaskExecutorRegistration;
+import io.mantisrx.server.worker.TaskExecutorGateway;
+import io.mantisrx.shaded.com.google.common.base.Throwables;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+class ResourceClusterAwareSchedulerActor extends AbstractActorWithTimers {
+
+    private final ResourceCluster resourceCluster;
+    private final ExecuteStageRequestFactory executeStageRequestFactory;
+    private final JobMessageRouter jobMessageRouter;
+    private final int maxScheduleRetries;
+    private final int maxCancelRetries;
+    private final Duration intervalBetweenRetries;
+
+    public static Props props(
+        int maxScheduleRetries,
+        int maxCancelRetries,
+        Duration intervalBetweenRetries,
+        final ResourceCluster resourceCluster,
+        final ExecuteStageRequestFactory executeStageRequestFactory,
+        final JobMessageRouter jobMessageRouter) {
+        return Props.create(ResourceClusterAwareSchedulerActor.class, maxScheduleRetries, maxCancelRetries, intervalBetweenRetries, resourceCluster, executeStageRequestFactory,
+            jobMessageRouter);
+    }
+
+    public ResourceClusterAwareSchedulerActor(
+        int maxScheduleRetries,
+        int maxCancelRetries,
+        Duration intervalBetweenRetries,
+        ResourceCluster resourceCluster,
+        ExecuteStageRequestFactory executeStageRequestFactory,
+        JobMessageRouter jobMessageRouter) {
+        this.resourceCluster = resourceCluster;
+        this.executeStageRequestFactory = executeStageRequestFactory;
+        this.jobMessageRouter = jobMessageRouter;
+        this.maxScheduleRetries = maxScheduleRetries;
+        this.intervalBetweenRetries = intervalBetweenRetries;
+        this.maxCancelRetries = maxCancelRetries;
+    }
+
+    @Override
+    public Receive createReceive() {
```

Comment:
what about when the scheduler is not master? should it stop scheduling requests etc..?

Reply:
if the scheduler is not master, then the default ServiceLifecycle shuts down the master and waits for it to get restarted.